File: java/src/main/java/org/openqa/selenium/support/ui/Select.java
package org.openqa.selenium.support.ui;

import java.util.ArrayList;
import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebElement;

/**
 * A {@code <select>} dropdown helper. Mirrors Selenium 4.x's
 * {@code org.openqa.selenium.support.ui.Select}: wraps a {@code <select>}
 * WebElement and drives it by finding and clicking its {@code <option>} children.
 *
 * <pre>{@code
 * new Select(driver.findElement(By.id("country"))).selectByVisibleText("Spain");
 * }</pre>
 */
public class Select {

    private final WebElement element;
    private final boolean isMulti;

    public Select(WebElement element) {
        String tag = element.getTagName().toLowerCase();
        if (!"select".equals(tag)) {
            throw new IllegalArgumentException(
                    "Select only works on <select> elements, not <" + tag + ">");
        }
        this.element = element;
        String multiple = element.getAttribute("multiple");
        this.isMulti = multiple != null && !"false".equals(multiple);
    }

    public WebElement getWrappedElement() {
        return element;
    }

    public boolean isMultiple() {
        return isMulti;
    }

    public List<WebElement> getOptions() {
        return element.findElements(By.tagName("option"));
    }

    public List<WebElement> getAllSelectedOptions() {
        List<WebElement> selected = new ArrayList<>();
        for (WebElement o : getOptions()) {
            if (o.isSelected()) {
                selected.add(o);
            }
        }
        return selected;
    }

    public WebElement getFirstSelectedOption() {
        for (WebElement o : getOptions()) {
            if (o.isSelected()) {
                return o;
            }
        }
        throw new NoSuchElementException("No options are selected", 17);
    }

    public void selectByVisibleText(String text) {
        for (WebElement o : getOptions()) {
            if (text.equals(o.getText())) {
                select(o);
                return;
            }
        }
        throw new NoSuchElementException("Cannot locate option with text: " + text, 17);
    }

    public void selectByValue(String value) {
        for (WebElement o : getOptions()) {
            if (value.equals(o.getAttribute("value"))) {
                select(o);
                return;
            }
        }
        throw new NoSuchElementException("Cannot locate option with value: " + value, 17);
    }

    public void selectByIndex(int index) {
        List<WebElement> opts = getOptions();
        if (index < 0 || index >= opts.size()) {
            throw new NoSuchElementException("Cannot locate option with index: " + index, 17);
        }
        select(opts.get(index));
    }

    public void deselectAll() {
        if (!isMulti) {
            throw new UnsupportedOperationException("You may only deselect all options of a multi-select");
        }
        for (WebElement o : getOptions()) {
            if (o.isSelected()) {
                o.click();
            }
        }
    }

    public void deselectByValue(String value) {
        requireMulti();
        boolean matched = false;
        for (WebElement o : getOptions()) {
            if (value.equals(o.getAttribute("value"))) {
                deselect(o);
                matched = true;
            }
        }
        if (!matched) {
            throw new NoSuchElementException("Cannot locate option with value: " + value, 17);
        }
    }

    public void deselectByIndex(int index) {
        requireMulti();
        for (WebElement o : getOptions()) {
            if (String.valueOf(index).equals(o.getAttribute("index"))) {
                deselect(o);
                return;
            }
        }
        throw new NoSuchElementException("Cannot locate option with index: " + index, 17);
    }

    public void deselectByVisibleText(String text) {
        requireMulti();
        boolean matched = false;
        for (WebElement o : getOptions()) {
            if (text.equals(o.getText())) {
                deselect(o);
                matched = true;
            }
        }
        if (!matched) {
            throw new NoSuchElementException("Cannot locate option with text: " + text, 17);
        }
    }

    private void requireMulti() {
        if (!isMulti) {
            throw new UnsupportedOperationException("You may only deselect options of a multi-select");
        }
    }

    private void select(WebElement option) {
        if (!option.isSelected()) {
            option.click();
        }
    }

    private void deselect(WebElement option) {
        if (option.isSelected()) {
            option.click();
        }
    }
}

File: python/selenium/webdriver/support/select.py
"""``<select>`` dropdown helper, at the Selenium 4.x import path
``from selenium.webdriver.support.select import Select``.

Wraps a ``<select>`` WebElement and drives it by finding and clicking its
``<option>`` children — the same approach mainstream Selenium's ``Select`` uses::

    from selenium.webdriver.support.select import Select
    Select(driver.find_element(By.ID, "country")).select_by_visible_text("Spain")
"""
from __future__ import annotations

from typing import List

from ..._webdriver import WebElement, NoSuchElementException
from ..common.by import By


class Select:
    def __init__(self, element: WebElement):
        tag = element.tag_name.lower()
        if tag != "select":
            raise ValueError(f"Select only works on <select> elements, not <{tag}>")
        self._el = element
        multi = element.get_attribute("multiple")
        self.is_multiple = bool(multi) and multi != "false"

    @property
    def options(self) -> List[WebElement]:
        return self._el.find_elements(By.TAG_NAME, "option")

    @property
    def all_selected_options(self) -> List[WebElement]:
        return [o for o in self.options if o.is_selected()]

    @property
    def first_selected_option(self) -> WebElement:
        for o in self.options:
            if o.is_selected():
                return o
        raise NoSuchElementException("no option is selected")

    def select_by_visible_text(self, text: str) -> None:
        for o in self.options:
            if o.text == text:
                self._select(o)
                return
        raise NoSuchElementException(f"no option with visible text {text!r}")

    def select_by_value(self, value: str) -> None:
        for o in self.options:
            if o.get_attribute("value") == value:
                self._select(o)
                return
        raise NoSuchElementException(f"no option with value {value!r}")

    def select_by_index(self, index: int) -> None:
        opts = self.options
        if index < 0 or index >= len(opts):
            raise NoSuchElementException(f"no option at index {index}")
        self._select(opts[index])

    def deselect_all(self) -> None:
        if not self.is_multiple:
            raise NotImplementedError("deselect_all only makes sense on a multi-select")
        for o in self.options:
            if o.is_selected():
                o.click()

    def _select(self, option: WebElement) -> None:
        if not option.is_selected():
            option.click()


__all__ = ["Select"]
